runtime/from-json-processor.h

```diff
@@ -0,0 +1,207 @@
+    do_set(key, value.ref(), json);
+  }
+
+  template<class I>
+  void do_set(std::string_view key, class_instance<I> &klass, const mixed &json) noexcept;
+
+  // just don't fail compilation with empty untyped arrays
+  void do_set(std::string_view /*key*/, array<Unknown> &/*array*/, const mixed &/*json*/) noexcept {}
+
+  template<class T>
+  void do_set_array(std::string_view key, array<T> &array, const mixed &json) noexcept {
+    const auto &json_array = json.as_array();
+    const auto array_size = json_array.size();
+    array.reserve(array_size.int_size, array_size.string_size, array_size.is_vector);
+
+    for (const auto it : json_array) {
+      auto json_key = it.get_key();
+      if (json_key.is_string() && json_key.as_string() == get_json_obj_magic_key()) {
+        continue; // don't deserialize magic
+      }
+      auto &elem = array[json_key]; // create value anyway despite that json value may be null
+      if (!it.get_value().is_null()) {
+        do_set(key, elem, it.get_value());
+      }
+    }
+  }
+
+  template<class T>
+  void do_set(std::string_view key, array<T> &array, const mixed &json) noexcept {
+    if (json.is_array()) {
+      do_set_array(key, array, json);
+    } else {
+      store_error_message_for(key, json);
+    }
+  }
+
+  void do_set(std::string_view key, mixed &value, const mixed &json) noexcept {
+    if (json.is_array()) {
+      array<mixed> array;
+      do_set(key, array, json);
+      value = std::move(array);
+    } else {
+      value = json;
+    }
+  }
+
+  string error_;
+  const mixed &json_;
+  bool flatten_class_{false};
+};
+
+template<class I, class Tag>
+class_instance<I> from_json_impl(const mixed &json) noexcept {
+  class_instance<I> instance;
+  if constexpr (std::is_empty_v<I>) {
+    instance.empty_alloc();
+  } else {
+    instance.alloc();
+    FromJsonVisitor<Tag> visitor{json, impl_::IsJsonFlattenClass<I>::value};
+    instance.get()->accept(visitor);
+    if (visitor.has_error()) {
+      JsonEncoderError::msg.append(visitor.get_error());
+      return {};
+    }
+  }
+  return JsonEncoderError::msg.empty() ? instance : class_instance<I>{};
+}
+
+template<class Tag>
+template<class I>
+void FromJsonVisitor<Tag>::do_set(std::string_view key, class_instance<I> &klass, const mixed &json) noexcept {
+  if constexpr (!impl_::IsJsonFlattenClass<I>::value) {
+    if (!json.is_array() || json.as_array().is_vector()) {
+      store_error_message_for(key, json);
+      return;
+    }
+  }
+  klass = from_json_impl<I, Tag>(json);
+}
+
+template<class ClassName, class Tag>
+ClassName f$from_json_impl(Tag /*tag*/, const string &json_string, const string &/*class_mame*/) noexcept {
+  JsonEncoderError::msg = {};
+
+  auto [json, success] = json_decode(json_string, FromJsonVisitor<Tag>::get_json_obj_magic_key().c_str());
+
+  if (!success) {
+    JsonEncoderError::msg.append("failed to parse json string");
```

Comment:
What about adding a separate error message for an empty string passed?
```
if (!success) {
  if (json_string.empty()) {
    JsonEncoderError::msg.append("passed an empty json string");
  } else {
    JsonEncoderError::msg.append("failed to parse json string");
  }
}
```

Reply:
I think not worth to deal with such rare case.